packages/zk-certificates/contracts/AirdropGateway.sol

```diff
@@ -0,0 +1,137 @@
+    mapping(uint => mapping(address => bool)) public registeredUsers;
+    // distribution Index -> humanId => registration status
+    mapping(uint => mapping(bytes32 => bool)) public registeredHumanID;
+    // distributionIndex -> userAddress -> claimed status
+    mapping(uint => mapping(address => bool)) public claimedUsers;
+    IZkKYCVerifier public verifierWrapper;
+
+    event DistributionCreated(uint indexed distributionId, address client);
+    event UserRegistered(uint indexed distributionId, address user);
+    event UserClaimed(uint indexed distributionId, address user, uint amount);
+
+    // struct to store airdrop distribution information
+    struct AirdropDistribution {
+        address[] requiredSBTs;
+        uint registrationStartTime;
+        uint registrationEndTime;
+        uint claimStartTime;
+        uint claimEndTime;
+        address clientAddress;
+        address tokenAddress;
+        uint256 distributionAmount;
+        uint256 registeredUserCount;
+        uint256 tokenAmountPerUser;
+        uint256 amountClaimed;
+    }
+
+    constructor(address owner, address verifierWrapperAddress) {
+        verifierWrapper = IZkKYCVerifier(verifierWrapperAddress);
+        // set admin, the role that can assign and revoke other roles
+        _setupRole(DEFAULT_ADMIN_ROLE, owner);
+    }
+
+    function whitelistClient(address newClient) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        _grantRole(CLIENT_ROLE, newClient);
+    }
+
+    function dewhitelistClient(address client) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        _revokeRole(CLIENT_ROLE, client);
+    }
+
+    function setDistribution(address[] memory requiredSBTs, address tokenAddress, uint registrationStartTime, uint registrationEndTime, uint claimStartTime, uint claimEndTime) external onlyRole(CLIENT_ROLE) {
+        require(registrationStartTime < registrationEndTime, "invalid registration time");
+        require(registrationEndTime < claimStartTime, "claim can only start after registration ends");
+        require(claimStartTime < claimEndTime, "invalid claim time");
+        distributions[distributionIndexCounter] = AirdropDistribution(requiredSBTs, registrationStartTime, registrationEndTime, claimStartTime, claimEndTime, msg.sender, tokenAddress, 0, 0, 0, 0);
+        emit DistributionCreated(distributionIndexCounter, msg.sender);
+        distributionIndexCounter++;
+    }
+
+    function getRequiredSBTs(uint distributionId) external view returns (address[] memory) {
+        return distributions[distributionId].requiredSBTs;
+    }
+
+    /* deposit tokens to the distribution
+     we choose to not sending the tokens to distribution contract address directly
+     to just in case distinguish between two distributions with the same token address
+    */
+    function deposit(uint distributionId, uint amount) external onlyRole(CLIENT_ROLE) {
+        IERC20(distributions[distributionId].tokenAddress).transferFrom(msg.sender, address(this), amount);
```

Comment:
Works, but I think the best practice guideline is to do asset transfers after the require checks.

Reply:
fixed